src/coffee/io/reader.coffee

```diff
@@ -0,0 +1,97 @@
+_ = require 'underscore'
+Csv = require 'csv'
+path = require 'path'
+Promise = require 'bluebird'
+iconv = require 'iconv-lite'
+fs = Promise.promisifyAll require('fs')
+Excel = require 'exceljs'
+
+DEBUG = false
+debugLog = if DEBUG then console.log else _.noop
+
+class Reader
+
+  constructor: (@options = {}) ->
+    debugLog "READER::options:", JSON.stringify(@options)
+    @options.encoding = @options.encoding || 'utf-8'
+    @header = null
+    @rows = []
+
+  read: (file) =>
+    # read from file or from stdin?
+    if file
+      debugLog "READER::stream file %s", file
+      @inputStream = fs.createReadStream file
+    else
+      debugLog "READER::stream stdin"
+      @inputStream = process.stdin
+
+    if @options.format == 'xlsx'
+      @_readXlsx(@inputStream)
+    else
+      @_readCsv(@inputStream)
+
+  @parseCsv: (csv, delimiter, encoding) ->
+    rows = []
+    options =
+      delimiter: delimiter
+      skip_empty_lines: true
+
+    # only buffer can be decoded from another encoding
+    if csv instanceof Buffer
+      csv = @decode(csv, encoding)
+
+    new Promise (resolve, reject) ->
+      Csv()
+      .from.string(csv, options)
+      .on 'record', (row) ->
+        rows.push(row)
+      .on 'error', (err) ->
+        reject(err)
+      .on 'end', () ->
+        resolve(rows)
+
+  _readCsv: (stream) =>
+    new Promise (resolve, reject) =>
+      buffers = []
+
+      # stream whole file to buffer because we need to decode it first from buffer
+      # - iconv-lite does not support string to string decoding
+      stream.on 'data', (buffer) ->
+        buffers.push buffer
```

Comment:
is there anyway to do this without storing all the data in memory?

Let's discuss.